Add single dimension -> multidimension partition mapping

clairelin135 opened the pull request:
Resolves https://github.com/dagster-io/dagster/issues/10722

Following https://github.com/dagster-io/dagster/pull/10909/files, this PR adds a partition mapping from an upstream single-dimensionally partitioned asset to a downstream multipartitioned asset. In order to use this class, the upstream partitions definition must be a dimension of the downstream asset.

The new `PartitionMapping` object also returns partition subsets instead of ranges, enabling use cases where a downstream subset is not a valid range e.g.:
```
MultiPartitionsDefinition(
    {"ab": StaticPartitionsDefinition(["a", "b"]), "12": StaticPartitionsDefinition(["1", "2"])}
)
```
has partition keys: ['1|a', '1|b', '2|a', '2|b']
If the upstream asset had partitions definition StaticPartitionsDefinition(["a", "b"]), the downstream partition keys subset should be {'1|a', '2|a'}, which can't be expressed as a range.

sryza commented on python_modules/dagster/dagster/_core/definitions/partition_mapping.py:
```diff
@@ -193,6 +197,145 @@ def get_downstream_partitions_for_partition_range(
 
 
+@experimental
+class SingleDimensionToMultiPartitionMapping(PartitionMapping):
```
From the name, I wasn't able to tell which direction this goes - i.e. whether the upstream or the downstream partitions definition is expected to be multi-. Is there a way to make that more clear, or at least conform to a convention?

FWIW `AllPartitionMapping` and `LastPartitionMapping`, prefer to what partitions in the upstream asset correspond to a partition of the downstream asset.

clairelin135 replied:
That's a good point. Renaming this to `SingleDimensionDependencyMapping`, hopefully that is clearer.

Tossing out some possible naming ideas:
- SingleDimensionDependencyMapping
- MatchesDownstreamPartitionDimensionMapping
- PartitionDimensionMapping

clairelin135 commented:
@sryza per your feedback I've renamed the class and switched to the new partition subset methods, raising `NotImplementedError`s for the range methods. I think this is ready for you to take another look!

sryza commented on python_modules/dagster/dagster/_core/definitions/partition_mapping.py:
```diff
@@ -215,6 +219,107 @@ def get_downstream_partitions_for_partition_range(
+            check.failed(f"Partition dimension '{self.partition_dimension_name}' not found")
+        matching_dimension_def = next(iter(matching_dimensions))
+
+        if upstream_partitions_def != matching_dimension_def.partitions_def:
```
Something to keep in mind for future optimization is that this `!=` could be expensive for large `StaticPartitionsDefinition`s.

sryza reviewed (approved):
Nice
